aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,639 @@
+
+    /// Resource Account Doesn't exist
+    const ERESOURCE_ACCT_NOT_EXIST: u64 = 16;
+
+    /// Proposal resource not created
+    const EPRPOSALS_NOT_EXIST_AT_ADDRESS: u64 = 17;
+
+    /// Proposal id doesn't exist
+    const EPRPOSAL_ID_NOT_EXIST: u64 = 18;
+
+
+    /// Constants
+    const PROPOSAL_UNRESOLVED: u8 = 0;
+    const PROPOSAL_RESOLVED_PASSED: u8 = 1;
+    const PROPOSAL_RESOLVED_NOT_PASSED: u8 = 2;
+
+    const DEFALT_VOTING_WEIGHT: u64 = 1;
+
+    struct DAO has key {
+        /// Name of the DAO
+        name: String,
+        /// The threshold that the proposal can resolve. For example, 50% means at least 50% of NFT holder should vote to resolve the proposal
+        resolve_threshold: ResolveThreshold,
+        /// The NFT Collection that is used to govern the DAO
+        governance_token: GovernanceToken, // This is the governance token or NFT used for voting
+        /// The address of the creator account this DAO owns. THis account is a resource account that this DAO is an adminstrator of the controlled account
+        controlled_account: address, // This is the nft creator account that this DAO governs, this accout should offer it signer cap to DAO account
+        /// The voting duration in secs
+        voting_duration: u64,
+        /// Total number of eligible member NFTs
+        total_supply: u64,
+        /// Minimum weight for proposal
+        min_proposal_weight: u64,
+        /// Proposal counter
+        cur_largest_proposal_id: u64,
+    }
+
+    struct ResolveThreshold has store {
+        numerator: u64,
+        denominator: u64,
+    }
+
+    /// The collection should only contains NFTs, where all token name only has 1 maximal and immutable
+    /// The total supply is fixed with the token names.
+    struct GovernanceToken has store {
+        /// The creator address of this NFT collection
+        creator: address,
+        /// The collection name
+        collection: String,
+        /// Eligible token_ids that can vote. Value of the table is the weights of the token
+        member_tokens: Table<TokenId, u64>
+    }
+
+    /// All proposals
+    struct Proposals has key {
+        proposals: Table<u64, Proposal>,
+    }
+
+    /// Store the general information about a proposal
+    struct Proposal has copy, drop, store {
+        /// Name of the proposal, limiting to 64 chars
+        name: String,
+        /// Description of the proposal, limiting to 512 chars
+        description: String,
+        /// The name of function to be executed
+        function_name: String,
+        /// The function arguments to be exectued
+        function_args: PropertyMap,
+        /// The start time of the voting
+        start_time_sec: u64,
+        /// Proposal results, unresolved, passed, not passed
+        resolution: u8,
+    }
+
+    struct ProposalVotingStatistics has key {
+        proposals: Table<u64, VotingStatistics>,
+    }
+
+    struct VotingStatistics has store {
```

Comment:
What are the benefits of having this separate from the proposal itself? I think the different steps below (vote, resolve) would still need to fetch both because of the start/end time stored in the proposal right?

Reply:
The VotingStatistics could consume a lot of storage. We would want to be able to destroy it once the proposal ends without destroying the proposal. The proposal here can be kept as an on-chain history of all the past results.

Reply:
I see it makes sense to destroy yes_votes and no_votes records but should total_yes and total_no be part of that permanent historical record then and be moved to proposal?